### ISSUE TITLE
(Merge button events with stylus tip events) does not work for Hand tool in the latest version 1.1.1

### ISSUE THREAD

**Issue by malnaanah** [xournalpp/xournalpp#3826]
Version 1.1.1. Linux, KDE (Kubuntu 20.04). appimage.

The hand tool is assigned to one of the stylus button and it does not work unless the stylus touch the pad. The "Merge button events with stylus tip events" is active and it was working in previous versions. 
The problem exists also in the latest 1.1.1+ nightly build.

**Comment by pobrn** [comment]
Seems to be caused by f961b04b42d7c5d0ddbb1e8f0b5785da0e9ffbdf.
```
git bisect start
# bad: [14e2186caccbe4ad441133821ea6bac526edd142] Properly handle blocking of devices for setsquare
git bisect bad 14e2186caccbe4ad441133821ea6bac526edd142
# good: [9d1277dee22bb095c2db047bb04f89cc837aee3c] [skip ci] Update translation template
git bisect good 9d1277dee22bb095c2db047bb04f89cc837aee3c
# good: [5917d3dbe98b6b8b601b11382555a802d675d6f1] Fixup of test not found by cmake
git bisect good 5917d3dbe98b6b8b601b11382555a802d675d6f1
# good: [8d9b722afed93bbb48c2b495ca10f5d1c6ac45d0] Minor code cleanup for VerticalToolHandler
git bisect good 8d9b722afed93bbb48c2b495ca10f5d1c6ac45d0
# bad: [bd01e8d6b578cf97aace4a2d3f9eef70043b5da1] [skip ci] Update translation template
git bisect bad bd01e8d6b578cf97aace4a2d3f9eef70043b5da1
# bad: [7258d0bec14c4414eb20032dc295b9d561534be8] Fix issues with release script
git bisect bad 7258d0bec14c4414eb20032dc295b9d561534be8
# bad: [ba0a547de7d29f6a5bbebaa82fa9907910b80089] Fixup GDK_PIXBUF_MODULE_FILE string replacement
git bisect bad ba0a547de7d29f6a5bbebaa82fa9907910b80089
# good: [c1ae6c31268d828c8c827fcedae749099e630c3c] Fix wrong version suffix in relevant files
git bisect good c1ae6c31268d828c8c827fcedae749099e630c3c
# bad: [57654c0105ba058699c7869d485f095940d17e3e] fix: Crash on startup with no recent files
git bisect bad 57654c0105ba058699c7869d485f095940d17e3e
# bad: [f961b04b42d7c5d0ddbb1e8f0b5785da0e9ffbdf] Fix tool activation
git bisect bad f961b04b42d7c5d0ddbb1e8f0b5785da0e9ffbdf
# first bad commit: [f961b04b42d7c5d0ddbb1e8f0b5785da0e9ffbdf] Fix tool activation
```

---

@rolandlo, @wtbgagoa, maybe something like this would work? This is arguably not the best way to fix it because there are edge cases here and there, but it seems to work mostly as expected.
```diff
diff --git a/src/core/gui/inputdevices/StylusInputHandler.cpp b/src/core/gui/inputdevices/StylusInputHandler.cpp
index ff5d21cf..a27c5fc8 100644
--- a/src/core/gui/inputdevices/StylusInputHandler.cpp
+++ b/src/core/gui/inputdevices/StylusInputHandler.cpp
@@ -146,6 +146,10 @@ void StylusInputHandler::setPressedState(InputEvent const& event) {
             default:
                 break;
         }
+
+        if (this->inputContext->getSettings()->getInputSystemTPCButtonEnabled()) {
+            this->deviceClassPressed |= this->modifier2 || this->modifier3;
+        }
     } else if (event.type == MOTION_EVENT) {
         if (this->eventsToIgnore == 0) {
             this->deviceClassPressed = true;
@@ -158,18 +162,16 @@ void StylusInputHandler::setPressedState(InputEvent const& event) {
                 break;
             case 2:
                 this->modifier2 = false;
-                if (this->inputContext->getSettings()->getInputSystemTPCButtonEnabled()) {
-                    this->deviceClassPressed = false;
-                }
                 break;
             case 3:
                 this->modifier3 = false;
-                if (this->inputContext->getSettings()->getInputSystemTPCButtonEnabled()) {
-                    this->deviceClassPressed = false;
-                }
             default:
                 break;
         }
+
+        if (this->inputContext->getSettings()->getInputSystemTPCButtonEnabled()) {
+            this->deviceClassPressed &= this->modifier2 || this->modifier3;
+        }
     }
 }
 
```

**Comment by rolandlo** [comment]
I have tested @pobrn's patch on X11 with TabletPCButton on and the "Merge button events with stylus tip events" option with all stylus buttons mapped to the hand tool. I have two buttons on the side of the stylus and one is the eraser button, the other a regular button. With both buttons I don't receive any input events while pressing the button and hovering. With the eraser button the (open) hand icon shows up when I press the button during the motion event. With the other button the hand icon doesn't show up. For both buttons the hand tool only works (and the icon is a closed hand) when the stylus tip touches the device. 

https://github.com/xournalpp/xournalpp/blob/16a151533f9a31be4c21a900ea112cece2a93171/src/core/gui/inputdevices/StylusInputHandler.cpp#L190-L202

Note that the patch only makes changes (and the `modifier2` and `modifier3` variables are only changed) on button press and button release events (which I don't receive with TabletPCButton on). So something must be different with the input events that @pobrn @malnaanah and @tolvanea receive. Currently I don't see what. Unfortunately I cannot test it on Windows currently, because my touchscreen of the Windows laptop is broken and I forgot to take my Wacom tablet with me. I can only test it on Sunday next week.

@pobrn @malnaanah @tolvanea If anyone of you could provide [input logs](https://github.com/xournalpp/xournalpp/wiki/How-to-provide-debug-info-for-input-related-bugs) that would help.

**Comment by rolandlo** [comment]
Ok, I see. If `TabletPCButton` is `off`, then "Merge button events with stylus tip events" are expected to be disabled in the settings. So I will check the case of `TabletPCButton` `off` combined with the "Merge" option enabled more closely.

**Comment by rolandlo** [comment]
Even when `TabletPCButton` is `off` and `Merge button events with stylus tip events` is enabled, I don't see any change (via https://github.com/xournalpp/xournalpp/commit/f961b04b42d7c5d0ddbb1e8f0b5785da0e9ffbdf or via @pobrn's patch) in behavior with respect to the hand tool working when pressing a stylus button while hovering:
- With the eraser button it doesn't work (since no button press and button release events are obtained)
- With the regular stylus button it does work

I tested both on X11 and wayland, no difference. So I really think we need input logs to solve that mistery.

**Comment by rolandlo** [comment]
You get the same input events like me when pressing the regular stylus button. Strangely for me the hand tool works while hovering in that scenario (`TabletPCButton=off` and `Merge button events with stylus tip events` enabled) and for you it doesn't.

**Comment by rolandlo** [comment]
I think I found the issue. Can you please check the "Ignore the first ... input events" checkbox (with any value, e.g. 1) in the Stylus tab of menu `Edit > Preferences`? Does it work then?

**Comment by rolandlo** [comment]
@bhennion Do you still think we should fix this issue for the 1.2 release? The desired behavior can be obtained by enabling the `Ignore the first ... inputs events` option along with the `Merge button events with stylus tip events` option and moreover this is more like a useful bug than a feature (when `TabletPCButton=off` one should not really enable the `Merge button events with stylus tip events` option, since that option is intended for devices that don't emit "long input sequences" like on Windows and on X11 with `TabletPCButton=on`).

**Comment by bhennion** [comment]
No I don't. We may even close this issue I guess.
